File: backend/app/parsers/comfyui.py
import json
from typing import Any, ClassVar

from app.parsers.base import (
    ControlNetInfo,
    LoraInfo,
    MetadataParser,
    ParsedMetadata,
    SourceTool,
)
from app.parsers.model_detector import detect_model_type


class ComfyUIParser(MetadataParser):
    """Parser for ComfyUI metadata format."""

    # Node types to extract data from
    SAMPLER_NODES: ClassVar[set[str]] = {
        "KSampler",
        "KSamplerAdvanced",
        "SamplerCustom",
    }
    CHECKPOINT_NODES: ClassVar[set[str]] = {
        "CheckpointLoaderSimple",
        "CheckpointLoader",
        "UNETLoader",
    }
    PROMPT_NODES: ClassVar[set[str]] = {"CLIPTextEncode", "CLIPTextEncodeSDXL"}
    LORA_NODES: ClassVar[set[str]] = {"LoraLoader", "LoraLoaderModelOnly"}
    CONTROLNET_NODES: ClassVar[set[str]] = {
        "ControlNetLoader",
        "ControlNetApply",
        "ControlNetApplyAdvanced",
    }
    UPSCALE_MODEL_NODES: ClassVar[set[str]] = {"UpscaleModelLoader"}
    UPSCALE_NODES: ClassVar[set[str]] = {
        "ImageUpscaleWithModel",
        "LatentUpscale",
        "LatentUpscaleBy",
        "ImageScaleBy",
    }

    def can_parse(self, png_info: dict[str, Any]) -> bool:
        """Check if the PNG info contains ComfyUI metadata."""
        if "prompt" not in png_info:
            return False

        try:
            prompt_data = png_info["prompt"]
            if isinstance(prompt_data, str):
                prompt_data = json.loads(prompt_data)

            # Check if it has the node structure
            if not isinstance(prompt_data, dict):
                return False

            # Check for at least one node with class_type
            for node in prompt_data.values():
                if isinstance(node, dict) and "class_type" in node:
                    return True

            return False
        except (json.JSONDecodeError, TypeError):
            return False

    def parse(self, png_info: dict[str, Any]) -> ParsedMetadata:
        """Parse ComfyUI metadata."""
        prompt_str = png_info.get("prompt", "{}")
        workflow_str = png_info.get("workflow", "{}")

        prompt_data = (
            json.loads(prompt_str) if isinstance(prompt_str, str) else prompt_str
        )
        workflow_data = (
            json.loads(workflow_str) if isinstance(workflow_str, str) else workflow_str
        )

        metadata = ParsedMetadata(
            source_tool=SourceTool.COMFYUI,
            has_metadata=True,
            raw_metadata={"prompt": prompt_data, "workflow": workflow_data},
        )

        # Extract data from nodes
        self._extract_sampler_data(prompt_data, metadata)
        self._extract_checkpoint_data(prompt_data, metadata)
        self._extract_prompts(prompt_data, metadata)
        self._extract_loras(prompt_data, metadata)
        self._extract_controlnets(prompt_data, metadata)
        self._extract_upscale_info(prompt_data, metadata)
        self._extract_workflow_extras(workflow_data, metadata)

        # Detect model type
        metadata.model_type = detect_model_type(metadata.model_name)

        return metadata

    def _extract_sampler_data(
        self, prompt_data: dict[str, Any], metadata: ParsedMetadata
    ) -> None:
        """Extract sampler parameters from KSampler nodes."""
        for node in prompt_data.values():
            if not isinstance(node, dict):
                continue

            class_type = node.get("class_type", "")
            if class_type not in self.SAMPLER_NODES:
                continue

            inputs = node.get("inputs", {})

            if "seed" in inputs:
                seed_val = inputs["seed"]
                if isinstance(seed_val, (int, float)):
                    metadata.seed = int(seed_val)

            if "steps" in inputs:
                steps_val = inputs["steps"]
                if isinstance(steps_val, (int, float)):
                    metadata.steps = int(steps_val)

            if "cfg" in inputs:
                cfg_val = inputs["cfg"]
                if isinstance(cfg_val, (int, float)):
                    metadata.cfg_scale = float(cfg_val)

            if "sampler_name" in inputs:
                metadata.sampler_name = str(inputs["sampler_name"])

            if "scheduler" in inputs:
                metadata.scheduler = str(inputs["scheduler"])

            # Found a sampler, stop looking
            break

    def _extract_checkpoint_data(
        self, prompt_data: dict[str, Any], metadata: ParsedMetadata
    ) -> None:
        """Extract checkpoint/model name."""
        for node in prompt_data.values():
            if not isinstance(node, dict):
                continue

            class_type = node.get("class_type", "")
            if class_type not in self.CHECKPOINT_NODES:
                continue

            inputs = node.get("inputs", {})

            ckpt_name = inputs.get("ckpt_name") or inputs.get("unet_name")
            if ckpt_name:
                # Remove file extension
                if isinstance(ckpt_name, str):
                    metadata.model_name = ckpt_name.rsplit(".", 1)[0]
                break

    def _extract_prompts(
        self, prompt_data: dict[str, Any], metadata: ParsedMetadata
    ) -> None:
        """Extract positive and negative prompts by following node references."""
        # Find KSampler to trace prompt references
        sampler_node = None
        for node in prompt_data.values():
            if isinstance(node, dict) and node.get("class_type") in self.SAMPLER_NODES:
                sampler_node = node
                break

        if not sampler_node:
            return

        inputs = sampler_node.get("inputs", {})

        # Get positive prompt
        positive_ref = inputs.get("positive")
        if positive_ref:
            positive_text = self._resolve_prompt_reference(prompt_data, positive_ref)
            if positive_text:
                metadata.positive_prompt = positive_text

        # Get negative prompt
        negative_ref = inputs.get("negative")
        if negative_ref:
            negative_text = self._resolve_prompt_reference(prompt_data, negative_ref)
            if negative_text:
                metadata.negative_prompt = negative_text

    def _resolve_prompt_reference(
        self, prompt_data: dict[str, Any], ref: Any
    ) -> str | None:
        """Resolve a node reference to get the prompt text."""
        if not isinstance(ref, list) or len(ref) < 1:
            return None

        node_id = str(ref[0])
        node = prompt_data.get(node_id)

        if not isinstance(node, dict):
            return None

        class_type = node.get("class_type", "")

        # Direct text encode node
        if class_type in self.PROMPT_NODES:
            inputs = node.get("inputs", {})
            text = inputs.get("text")
            if isinstance(text, str):
                return text
            # SDXL uses text_g and text_l
            text_g = inputs.get("text_g")
            if isinstance(text_g, str):
                return text_g

        # Might be a conditioning node, try to follow the chain
        inputs = node.get("inputs", {})
        for key in ["conditioning", "clip"]:
            next_ref = inputs.get(key)
            if isinstance(next_ref, list):
                result = self._resolve_prompt_reference(prompt_data, next_ref)
                if result:
                    return result

        return None

    def _extract_loras(
        self, prompt_data: dict[str, Any], metadata: ParsedMetadata
    ) -> None:
        """Extract LoRA information."""
        for node in prompt_data.values():
            if not isinstance(node, dict):
                continue

            class_type = node.get("class_type", "")
            if class_type not in self.LORA_NODES:
                continue

            inputs = node.get("inputs", {})
            lora_name = inputs.get("lora_name")

            if lora_name:
                weight = inputs.get("strength_model", 1.0)
                weight_clip = inputs.get("strength_clip")

                if isinstance(lora_name, str):
                    lora_info = LoraInfo(
                        name=lora_name.rsplit(".", 1)[0],
                        weight=(
                            float(weight) if isinstance(weight, (int, float)) else 1.0
                        ),
                        weight_clip=(
                            float(weight_clip)
                            if isinstance(weight_clip, (int, float))
                            else None
                        ),
                    )
                    metadata.loras.append(lora_info)

    def _extract_controlnets(
        self, prompt_data: dict[str, Any], metadata: ParsedMetadata
    ) -> None:
        """Extract ControlNet information."""
        controlnet_models: dict[str, str] = {}

        # First pass: collect model names
        for node_id, node in prompt_data.items():
            if not isinstance(node, dict):
                continue

            if node.get("class_type") == "ControlNetLoader":
                inputs = node.get("inputs", {})
                model_name = inputs.get("control_net_name")
                if model_name:
                    controlnet_models[node_id] = model_name

        # Second pass: collect apply nodes with weights
        for node in prompt_data.values():
            if not isinstance(node, dict):
                continue

            class_type = node.get("class_type", "")
            if class_type not in {"ControlNetApply", "ControlNetApplyAdvanced"}:
                continue

            inputs = node.get("inputs", {})
            strength = inputs.get("strength", 1.0)

            # Try to find the model reference
            control_net_ref = inputs.get("control_net")
            model_name = "unknown"

            if isinstance(control_net_ref, list) and len(control_net_ref) >= 1:
                loader_id = str(control_net_ref[0])
                model_name = controlnet_models.get(loader_id, "unknown")

            cn_info = ControlNetInfo(
                model=model_name,
                weight=float(strength) if isinstance(strength, (int, float)) else 1.0,
                guidance_start=float(inputs.get("start_percent", 0.0)),
                guidance_end=float(inputs.get("end_percent", 1.0)),
            )
            metadata.controlnets.append(cn_info)

    def _extract_upscale_info(
        self, prompt_data: dict[str, Any], metadata: ParsedMetadata
    ) -> None:
        """Extract upscale/hires information."""
        upscale_models: dict[str, str] = {}

        # First pass: collect upscale model names
        for node_id, node in prompt_data.items():
            if not isinstance(node, dict):
                continue

            if node.get("class_type") in self.UPSCALE_MODEL_NODES:
                inputs = node.get("inputs", {})
                model_name = inputs.get("model_name")
                if model_name:
                    upscale_models[node_id] = model_name

        # Second pass: detect upscale usage
        for node in prompt_data.values():
            if not isinstance(node, dict):
                continue

            class_type = node.get("class_type", "")
            inputs = node.get("inputs", {})

            if class_type == "ImageUpscaleWithModel":
                # Find the model reference
                upscale_model_ref = inputs.get("upscale_model")
                model_name = "unknown"
                if isinstance(upscale_model_ref, list) and len(upscale_model_ref) >= 1:
                    loader_id = str(upscale_model_ref[0])
                    model_name = upscale_models.get(loader_id, "unknown")

                metadata.model_params["hires_upscaler"] = model_name
                metadata.model_params["upscale_method"] = "model"

            elif class_type == "LatentUpscale":
                method = inputs.get("upscale_method", "nearest-exact")
                width = inputs.get("width")
                height = inputs.get("height")
                metadata.model_params["hires_upscaler"] = f"Latent ({method})"
                metadata.model_params["upscale_method"] = "latent"
                if width and height:
                    metadata.model_params["upscale_size"] = f"{width}x{height}"

            elif class_type == "LatentUpscaleBy":
                method = inputs.get("upscale_method", "nearest-exact")
                scale = inputs.get("scale_by", 1.0)
                metadata.model_params["hires_upscaler"] = f"Latent ({method})"
                metadata.model_params["hires_upscale"] = (
                    float(scale) if isinstance(scale, (int, float)) else 1.0
                )
                metadata.model_params["upscale_method"] = "latent"

            elif class_type == "ImageScaleBy":
                method = inputs.get("upscale_method", "nearest-exact")
                scale = inputs.get("scale_by", 1.0)
                metadata.model_params["hires_upscaler"] = f"Image ({method})"
                metadata.model_params["hires_upscale"] = (
                    float(scale) if isinstance(scale, (int, float)) else 1.0
                )
                metadata.model_params["upscale_method"] = "image"

    def _extract_workflow_extras(
        self, workflow_data: dict[str, Any], metadata: ParsedMetadata
    ) -> None:
        """Extract additional workflow information."""
        if not workflow_data:
            return

        extras: dict[str, Any] = {}

        # Count nodes
        nodes = workflow_data.get("nodes", [])
        if isinstance(nodes, list):
            extras["node_count"] = len(nodes)

        # Get workflow version if available
        version = workflow_data.get("version")
        if version:
            extras["workflow_version"] = version

        if extras:
            metadata.workflow_extras = extras

File: backend-rs/src/parser/comfyui.rs
//! ComfyUI workflow-graph parser (mirror parsers/comfyui.py).
//!
//! The PNG `prompt` chunk is a JSON object mapping node-id → {class_type,
//! inputs}. Values reference other nodes as `[node_id, output_index]`. We walk
//! the graph to pull out sampler settings, checkpoint, prompts, loras,
//! controlnets and upscale info. Node iteration order follows insertion order
//! (serde_json `preserve_order`), matching the Python dict behaviour.

use std::collections::HashMap;

use serde_json::{Map, Value};

use super::{detect_model_type, ControlNetInfo, LoraInfo, ParsedMetadata, SourceTool};

const SAMPLER_NODES: &[&str] = &["KSampler", "KSamplerAdvanced", "SamplerCustom"];
const CHECKPOINT_NODES: &[&str] = &["CheckpointLoaderSimple", "CheckpointLoader", "UNETLoader"];
const PROMPT_NODES: &[&str] = &["CLIPTextEncode", "CLIPTextEncodeSDXL"];
const LORA_NODES: &[&str] = &["LoraLoader", "LoraLoaderModelOnly"];
const UPSCALE_MODEL_NODES: &[&str] = &["UpscaleModelLoader"];

/// Guard against cyclic graphs when following prompt references.
const MAX_REF_DEPTH: u32 = 30;

pub fn can_parse(png_info: &HashMap<String, String>) -> bool {
    let Some(prompt) = png_info.get("prompt") else {
        return false;
    };
    let Ok(data) = serde_json::from_str::<Value>(prompt) else {
        return false;
    };
    let Some(obj) = data.as_object() else {
        return false;
    };
    obj.values()
        .any(|n| n.is_object() && n.get("class_type").is_some())
}

pub fn parse(png_info: &HashMap<String, String>) -> ParsedMetadata {
    let prompt_data: Value = png_info
        .get("prompt")
        .and_then(|s| serde_json::from_str(s).ok())
        .unwrap_or_else(|| Value::Object(Map::new()));
    let workflow_data: Value = png_info
        .get("workflow")
        .and_then(|s| serde_json::from_str(s).ok())
        .unwrap_or_else(|| Value::Object(Map::new()));

    let mut m = ParsedMetadata::new(SourceTool::Comfyui);
    m.raw_metadata = Some(serde_json::json!({
        "prompt": prompt_data.clone(),
        "workflow": workflow_data.clone(),
    }));

    let empty = Map::new();
    let prompt = prompt_data.as_object().unwrap_or(&empty);

    extract_sampler_data(prompt, &mut m);
    extract_checkpoint_data(prompt, &mut m);
    extract_prompts(prompt, &mut m);
    extract_loras(prompt, &mut m);
    extract_controlnets(prompt, &mut m);
    extract_upscale_info(prompt, &mut m);
    extract_workflow_extras(&workflow_data, &mut m);

    m.model_type = Some(detect_model_type(m.model_name.as_deref()));
    m
}

fn class_type(node: &Value) -> &str {
    node.get("class_type").and_then(Value::as_str).unwrap_or("")
}

fn inputs(node: &Value) -> Option<&Map<String, Value>> {
    node.get("inputs").and_then(Value::as_object)
}

/// JSON number → i64 (accepts integral floats), excluding bools.
fn num_i64(v: &Value) -> Option<i64> {
    v.as_i64().or_else(|| v.as_f64().map(|f| f as i64))
}

fn num_f64(v: &Value) -> Option<f64> {
    if v.is_boolean() {
        return None;
    }
    v.as_f64()
}

/// Strip a single trailing file extension (mirror `name.rsplit(".", 1)[0]`).
fn strip_ext(name: &str) -> String {
    match name.rsplit_once('.') {
        Some((stem, _)) => stem.to_string(),
        None => name.to_string(),
    }
}

fn extract_sampler_data(prompt: &Map<String, Value>, m: &mut ParsedMetadata) {
    for node in prompt.values() {
        if !node.is_object() || !SAMPLER_NODES.contains(&class_type(node)) {
            continue;
        }
        if let Some(inp) = inputs(node) {
            if let Some(v) = inp.get("seed").and_then(num_i64) {
                m.seed = Some(v);
            }
            if let Some(v) = inp.get("steps").and_then(num_i64) {
                m.steps = Some(v as i32);
            }
            if let Some(v) = inp.get("cfg").and_then(num_f64) {
                m.cfg_scale = Some(v);
            }
            if let Some(v) = inp.get("sampler_name").and_then(Value::as_str) {
                m.sampler_name = Some(v.to_string());
            }
            if let Some(v) = inp.get("scheduler").and_then(Value::as_str) {
                m.scheduler = Some(v.to_string());
            }
        }
        break;
    }
}

fn extract_checkpoint_data(prompt: &Map<String, Value>, m: &mut ParsedMetadata) {
    for node in prompt.values() {
        if !node.is_object() || !CHECKPOINT_NODES.contains(&class_type(node)) {
            continue;
        }
        if let Some(inp) = inputs(node) {
            let ckpt = inp
                .get("ckpt_name")
                .and_then(Value::as_str)
                .or_else(|| inp.get("unet_name").and_then(Value::as_str));
            if let Some(name) = ckpt {
                m.model_name = Some(strip_ext(name));
                break;
            }
        }
    }
}

fn extract_prompts(prompt: &Map<String, Value>, m: &mut ParsedMetadata) {
    let Some(sampler) = prompt
        .values()
        .find(|n| n.is_object() && SAMPLER_NODES.contains(&class_type(n)))
    else {
        return;
    };
    let Some(inp) = inputs(sampler) else {
        return;
    };

    if let Some(pos) = inp.get("positive")
        && let Some(text) = resolve_prompt_reference(prompt, pos, MAX_REF_DEPTH)
    {
        m.positive_prompt = Some(text);
    }
    if let Some(neg) = inp.get("negative")
        && let Some(text) = resolve_prompt_reference(prompt, neg, MAX_REF_DEPTH)
    {
        m.negative_prompt = Some(text);
    }
}

fn ref_node_id(r: &Value) -> Option<String> {
    match r.as_array()?.first()? {
        Value::String(s) => Some(s.clone()),
        Value::Number(n) => Some(n.to_string()),
        _ => None,
    }
}

fn resolve_prompt_reference(prompt: &Map<String, Value>, r: &Value, depth: u32) -> Option<String> {
    if depth == 0 {
        return None;
    }
    let node = prompt.get(&ref_node_id(r)?)?;
    if !node.is_object() {
        return None;
    }
    let inp = inputs(node);

    if PROMPT_NODES.contains(&class_type(node))
        && let Some(inp) = inp
    {
        if let Some(t) = inp.get("text").and_then(Value::as_str) {
            return Some(t.to_string());
        }
        if let Some(tg) = inp.get("text_g").and_then(Value::as_str) {
            return Some(tg.to_string());
        }
    }

    // Follow conditioning/clip links toward the text-encode node.
    if let Some(inp) = inp {
        for key in ["conditioning", "clip"] {
            if let Some(next) = inp.get(key)
                && next.is_array()
                && let Some(result) = resolve_prompt_reference(prompt, next, depth - 1)
            {
                return Some(result);
            }
        }
    }
    None
}

fn extract_loras(prompt: &Map<String, Value>, m: &mut ParsedMetadata) {
    for node in prompt.values() {
        if !node.is_object() || !LORA_NODES.contains(&class_type(node)) {
            continue;
        }
        let Some(inp) = inputs(node) else { continue };
        let Some(name) = inp.get("lora_name").and_then(Value::as_str) else {
            continue;
        };
        let weight = inp.get("strength_model").and_then(num_f64).unwrap_or(1.0);
        let weight_clip = inp.get("strength_clip").and_then(num_f64);
        m.loras.push(LoraInfo {
            name: strip_ext(name),
            weight,
            weight_clip,
            hash: None,
        });
    }
}

fn extract_controlnets(prompt: &Map<String, Value>, m: &mut ParsedMetadata) {
    // Pass 1: ControlNetLoader node-id → model name.
    let mut models: HashMap<String, String> = HashMap::new();
    for (node_id, node) in prompt {
        if node.is_object() && class_type(node) == "ControlNetLoader"
            && let Some(name) = inputs(node)
                .and_then(|i| i.get("control_net_name"))
                .and_then(Value::as_str)
        {
            models.insert(node_id.clone(), name.to_string());
        }
    }

    // Pass 2: Apply nodes carry the strength and reference a loader.
    for node in prompt.values() {
        if !node.is_object()
            || !matches!(class_type(node), "ControlNetApply" | "ControlNetApplyAdvanced")
        {
            continue;
        }
        let Some(inp) = inputs(node) else { continue };
        let strength = inp.get("strength").and_then(num_f64).unwrap_or(1.0);
        let model = inp
            .get("control_net")
            .and_then(ref_node_id)
            .and_then(|id| models.get(&id).cloned())
            .unwrap_or_else(|| "unknown".to_string());
        m.controlnets.push(ControlNetInfo {
            model,
            weight: strength,
            guidance_start: inp.get("start_percent").and_then(num_f64).unwrap_or(0.0),
            guidance_end: inp.get("end_percent").and_then(num_f64).unwrap_or(1.0),
            preprocessor: None,
        });
    }
}

fn extract_upscale_info(prompt: &Map<String, Value>, m: &mut ParsedMetadata) {
    let mut models: HashMap<String, String> = HashMap::new();
    for (node_id, node) in prompt {
        if node.is_object() && UPSCALE_MODEL_NODES.contains(&class_type(node))
            && let Some(name) = inputs(node)
                .and_then(|i| i.get("model_name"))
                .and_then(Value::as_str)
        {
            models.insert(node_id.clone(), name.to_string());
        }
    }

    for node in prompt.values() {
        if !node.is_object() {
            continue;
        }
        let Some(inp) = inputs(node) else { continue };
        let mp = &mut m.model_params;
        match class_type(node) {
            "ImageUpscaleWithModel" => {
                let model = inp
                    .get("upscale_model")
                    .and_then(ref_node_id)
                    .and_then(|id| models.get(&id).cloned())
                    .unwrap_or_else(|| "unknown".to_string());
                mp.insert("hires_upscaler".into(), Value::String(model));
                mp.insert("upscale_method".into(), Value::String("model".into()));
            }
            "LatentUpscale" => {
                let method = inp
                    .get("upscale_method")
                    .and_then(Value::as_str)
                    .unwrap_or("nearest-exact");
                mp.insert("hires_upscaler".into(), Value::String(format!("Latent ({method})")));
                mp.insert("upscale_method".into(), Value::String("latent".into()));
                let width = inp.get("width").and_then(num_i64);
                let height = inp.get("height").and_then(num_i64);
                if let (Some(w), Some(h)) = (width, height) {
                    mp.insert("upscale_size".into(), Value::String(format!("{w}x{h}")));
                }
            }
            "LatentUpscaleBy" => {
                let method = inp
                    .get("upscale_method")
                    .and_then(Value::as_str)
                    .unwrap_or("nearest-exact");
                let scale = inp.get("scale_by").and_then(num_f64).unwrap_or(1.0);
                mp.insert("hires_upscaler".into(), Value::String(format!("Latent ({method})")));
                mp.insert("hires_upscale".into(), serde_json::json!(scale));
                mp.insert("upscale_method".into(), Value::String("latent".into()));
            }
            "ImageScaleBy" => {
                let method = inp
                    .get("upscale_method")
                    .and_then(Value::as_str)
                    .unwrap_or("nearest-exact");
                let scale = inp.get("scale_by").and_then(num_f64).unwrap_or(1.0);
                mp.insert("hires_upscaler".into(), Value::String(format!("Image ({method})")));
                mp.insert("hires_upscale".into(), serde_json::json!(scale));
                mp.insert("upscale_method".into(), Value::String("image".into()));
            }
            _ => {}
        }
    }
}

fn extract_workflow_extras(workflow: &Value, m: &mut ParsedMetadata) {
    let Some(obj) = workflow.as_object() else {
        return;
    };
    if let Some(nodes) = obj.get("nodes").and_then(Value::as_array) {
        m.workflow_extras
            .insert("node_count".into(), serde_json::json!(nodes.len()));
    }
    if let Some(version) = obj.get("version")
        && !version.is_null()
    {
        m.workflow_extras
            .insert("workflow_version".into(), version.clone());
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_basic_comfyui_graph() {
        let prompt = r#"{
            "3": {"class_type": "KSampler", "inputs": {
                "seed": 42, "steps": 20, "cfg": 7.5,
                "sampler_name": "euler", "scheduler": "normal",
                "positive": ["6", 0], "negative": ["7", 0]
            }},
            "4": {"class_type": "CheckpointLoaderSimple", "inputs": {"ckpt_name": "animagine.safetensors"}},
            "6": {"class_type": "CLIPTextEncode", "inputs": {"text": "a cat"}},
            "7": {"class_type": "CLIPTextEncode", "inputs": {"text": "bad"}}
        }"#;
        let mut info = HashMap::new();
        info.insert("prompt".to_string(), prompt.to_string());
        assert!(can_parse(&info));
        let m = parse(&info);
        assert_eq!(m.source_tool, SourceTool::Comfyui);
        assert_eq!(m.seed, Some(42));
        assert_eq!(m.steps, Some(20));
        assert_eq!(m.cfg_scale, Some(7.5));
        assert_eq!(m.sampler_name.as_deref(), Some("euler"));
        assert_eq!(m.scheduler.as_deref(), Some("normal"));
        assert_eq!(m.model_name.as_deref(), Some("animagine"));
        assert_eq!(m.positive_prompt.as_deref(), Some("a cat"));
        assert_eq!(m.negative_prompt.as_deref(), Some("bad"));
    }
}
